source/extensions/filters/network/thrift_proxy/binary_protocol.cc

```diff
@@ -0,0 +1,300 @@
+#include "extensions/filters/network/thrift_proxy/binary_protocol.h"
+
+#include "envoy/common/exception.h"
+
+#include "common/common/assert.h"
+#include "common/common/fmt.h"
+#include "common/common/macros.h"
+
+#include "extensions/filters/network/thrift_proxy/buffer_helper.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ThriftProxy {
+
+const uint16_t BinaryProtocolImpl::Magic = 0x8001;
+
+bool BinaryProtocolImpl::readMessageBegin(Buffer::Instance& buffer, std::string& name,
+                                          MessageType& msg_type, int32_t& seq_id) {
+  // Minimum message length:
+  //   version: 2 bytes +
+  //   unused: 1 byte +
+  //   msg type: 1 byte +
+  //   name len: 4 bytes +
+  //   name: 0 bytes +
+  //   seq id: 4 bytes
+  if (buffer.length() < 12) {
+    return false;
+  }
+
+  uint16_t version = BufferHelper::peekU16(buffer);
+  if (version != Magic) {
+    throw EnvoyException(
+        fmt::format("invalid binary protocol version 0x{:04x} != 0x{:04x}", version, Magic));
+  }
+
+  // The byte at offset 2 is unused and ignored.
+
+  MessageType type = static_cast<MessageType>(BufferHelper::peekI8(buffer, 3));
+  if (type < MessageType::Call || type > MessageType::LastMessageType) {
+    throw EnvoyException(
+        fmt::format("invalid binary protocol message type {}", static_cast<int8_t>(type)));
+  }
+
+  uint32_t name_len = BufferHelper::peekU32(buffer, 4);
+  if (buffer.length() < name_len + 12) {
+    return false;
+  }
+
+  buffer.drain(8);
+
+  if (name_len > 0) {
+    name.assign(std::string(static_cast<char*>(buffer.linearize(name_len)), name_len));
+    buffer.drain(name_len);
+  } else {
+    name.clear();
+  }
+  msg_type = type;
+  seq_id = BufferHelper::drainI32(buffer);
+
+  return true;
+}
+
+bool BinaryProtocolImpl::readMessageEnd(Buffer::Instance& buffer) {
+  UNREFERENCED_PARAMETER(buffer);
+  return true;
+}
+
+bool BinaryProtocolImpl::readStructBegin(Buffer::Instance& buffer, std::string& name) {
+  UNREFERENCED_PARAMETER(buffer);
+  name.clear(); // binary protocol does not transmit struct names
+  return true;
+}
+
+bool BinaryProtocolImpl::readStructEnd(Buffer::Instance& buffer) {
+  UNREFERENCED_PARAMETER(buffer);
+  return true;
+}
+
+bool BinaryProtocolImpl::readFieldBegin(Buffer::Instance& buffer, std::string& name,
+                                        FieldType& field_type, int16_t& field_id) {
+  // FieldType::Stop is encoded as 1 byte.
+  if (buffer.length() < 1) {
+    return false;
+  }
+
+  FieldType type = static_cast<FieldType>(BufferHelper::peekI8(buffer));
+  if (type == FieldType::Stop) {
+    field_id = 0;
+    buffer.drain(1);
+  } else {
+    // FieldType followed by 2 bytes of field id
+    if (buffer.length() < 3) {
+      return false;
+    }
+    field_id = BufferHelper::peekI16(buffer, 1);
```

Comment:
Is it valid for the field ID to be negative?

Reply:
The protocol specifications all refer to it as a signed integer with no mention of the validity of negative values.

The Thrift IDL documents make no mention of the range of values allowed. In fact, the grammar defines it as an IntConstant, which implies negative values are legal. Empirically, the Thrift compiler will only accept values from 1 to 32767.

In practice, the generated bindings allow negative field IDs, but ignore them like any other unknown field ID.

Reply:
I'm going to leave this as-is from that standpoint that it's better to be a bit lenient in what's accepted.